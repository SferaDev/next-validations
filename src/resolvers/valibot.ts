import { parse } from 'valibot'
import type v from 'valibot'

export function valibotResolver<T extends v.AnySchema>(schema: T) {
  return {
    validate(data: unknown) {
      return parse(schema, data)
    },
  };
}
