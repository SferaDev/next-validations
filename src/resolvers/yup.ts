import * as Yup from 'yup';

export function yupResolver<T extends Yup.AnyObjectSchema>(schema: T) {
  return {
    validate: (data: unknown) => schema.validateSync(data),
  };
}
