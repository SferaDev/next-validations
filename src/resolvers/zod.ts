import { type z } from 'zod';

export function zodResolver<T extends z.ZodAny>(schema: T) {
  return {
    validate: (data: unknown) => schema.parse(data),
  };
}
