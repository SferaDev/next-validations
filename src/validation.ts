import type { LiteralUnion } from "type-fest";

import { fastestValidatorResolver } from "./resolvers/fastest-validator";
import { joiResolver } from "./resolvers/joi";
import { yupResolver } from "./resolvers/yup";
import { zodResolver } from "./resolvers/zod";
import { valibotResolver } from "./resolvers/valibot";

export type SCHEMA_TYPE = LiteralUnion<
  | "Yup"
  | "FastestValidator"
  | "Joi"
  | "Zod",
  | "Valibot"
  | string
>;

export function createResolver(type: SCHEMA_TYPE, schema: any) {
  switch (type) {
    case "Yup":
      return yupResolver(schema);
    case "FastestValidator":
      return fastestValidatorResolver(schema);
    case "Joi":
      return joiResolver(schema);
    case "Zod":
      return zodResolver(schema);
    case "Valibot":
      return valibotResolver(schema);


    default:
      throw new Error(`Does not support ${type} validation yet!`);
  }
}
