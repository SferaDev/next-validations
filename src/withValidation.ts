import { type NextApiRequest, type NextApiResponse } from "next";

import { createResolver } from "./validation";
import type { SCHEMA_TYPE } from "./validation";

type NextHandler = (err?: Error) => void;

type ValidationHoF = {
  type: SCHEMA_TYPE;
  mode?: "body" | "query" | "headers";
  schema: unknown;
};

export function withValidation({
  type,
  schema,
  mode = "query",
}: ValidationHoF) {
  return withValidations([{ type, schema, mode }]);
}

export function withValidations(validations: ValidationHoF[]) {
  return (
    handler?: (req: NextApiRequest, res: NextApiResponse) => any,
  ) => async (
    req: NextApiRequest,
    res: NextApiResponse,
    next?: NextHandler,
  ) => {
      try {
        validations.forEach((validation) => {
          const resolver = createResolver(validation.type, validation.schema);
          resolver.validate(req[validation.mode ?? "query"]);
        });

        if (next) {
          next(); return;
        }

        if (handler) return handler(req, res);

        res.status(404).end();
      } catch (error) {
        res.status(400).send(error);
      }
    };
}
