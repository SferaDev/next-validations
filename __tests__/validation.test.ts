import Joi from 'joi';
import * as valibot from 'valibot';
import * as yup from 'yup';
import { z } from 'zod';

import { createResolver } from '../src/validation';

describe('Yup Validation', () => {
  it('should create yup resolve base on schema', () => {
    const resolver = createResolver('Yup', {});
    expect(resolver).toMatchSnapshot();
  });

  it('should return true', () => {
    const schema = yup.object().shape({
      name: yup.string().required(),
      age: yup
        .number()
        .required()
        .positive()
        .integer(),
      email: yup.string().email(),
      website: yup.string().url(),
      createdOn: yup.date().default(function() {
        return new Date();
      }),
    });
    const resolver = createResolver('Yup', schema);
    const isValid = resolver.validate({
      name: 'jimmy',
      age: 24,
    });
    expect(isValid).toBeTruthy();
  });

  it('should return false', () => {
    const schema = yup.object().shape({
      name: yup.string().required(),
      age: yup
        .number()
        .required()
        .positive()
        .integer(),
      email: yup.string().email(),
      website: yup.string().url(),
      createdOn: yup.date().default(function() {
        return new Date();
      }),
    });
    const resolver = createResolver('Yup', schema);
    try {
      resolver.validate({
        name: 'jimmy',
      });
    } catch (error) {
      expect(error).toMatchSnapshot();
    }
  });
});

describe('FastestValidator Validation', () => {
  it('should create resolve base on schema', () => {
    const resolver = createResolver('FastestValidator', {});
    expect(resolver).toMatchSnapshot();
  });

  it('should return true', () => {
    const schema = {
      id: { type: 'number', positive: true, integer: true },
      name: { type: 'string', min: 3, max: 255 },
      status: 'boolean', // short-hand def
    };
    const resolver = createResolver('FastestValidator', schema);
    const isValid = resolver.validate({
      id: 5,
      name: 'John',
      status: true,
    });
    expect(isValid).toBeTruthy();
  });

  it('should return false', () => {
    const schema = {
      id: { type: 'number', positive: true, integer: true },
      name: { type: 'string', min: 3, max: 255 },
      status: 'boolean', // short-hand def
    };
    const resolver = createResolver('FastestValidator', schema);
    try {
      resolver.validate({
        id: 2,
        name: 'Adam',
      });
    } catch (error) {
      expect(error).toMatchSnapshot();
    }
  });
});

describe('Yoi Validation', () => {
  it('should create yoi resolve base on schema', () => {
    try {
      createResolver('Joi', {});
    } catch (error) {
      expect(error).toMatchSnapshot();
    }
  });

  it('should return true', () => {
    const schema = Joi.object({
      dob: Joi.date().iso(),
      email: Joi.string()
        .email()
        .required(),
      name: Joi.string().required(),
    });

    const resolver = createResolver('Joi', schema);
    const isValid = resolver.validate({
      name: 'Huynh Duc Dung',
      email: 'dung@productsway.com',
      dob: 1988,
    });
    expect(isValid).toBeTruthy();
  });

  it('should return false', () => {
    const schema = Joi.object({
      dob: Joi.date().iso(),
      email: Joi.string()
        .email()
        .required(),
      name: Joi.string().required(),
    });
    const resolver = createResolver('Joi', schema);
    try {
      resolver.validate({});
    } catch (error) {
      expect(error).toMatchSnapshot();
    }
  });
});

describe('Zod Validation', () => {
  it('should create zod resolve base on schema', () => {
    try {
      createResolver('Zod', {});
    } catch (error) {
      expect(error).toMatchSnapshot();
    }
  });

  it('should return true', () => {
    const schema = z.object({
      username: z.string(),
    });

    const resolver = createResolver('Zod', schema);
    const isValid = resolver.validate({
      username: 'jellydn',
    });
    expect(isValid).toBeTruthy();
  });

  it('should return false', () => {
    const schema = z.object({
      username: z.string().min(8),
    });
    const resolver = createResolver('Zod', schema);
    try {
      resolver.validate({
        username: 'jellydn',
      });
    } catch (error) {
      expect(error).toMatchSnapshot();
    }
  });
});

describe('Valibot Validation', () => {
  it('should create zod resolve base on schema', () => {
    try {
      createResolver('Valibot', {});
    } catch (error) {
      expect(error).toMatchSnapshot();
    }
  });

  it('should return true', () => {
    const schema = valibot.object({
      username: valibot.string(),
    });

    const resolver = createResolver('Valibot', schema);
    const isValid = resolver.validate({
      username: 'jellydn',
    });
    expect(isValid).toBeTruthy();
  });

  it('should return false', () => {
    const schema = valibot.object({
      age: valibot.number()
    });
    const resolver = createResolver('Valibot', schema);
    try {
      resolver.validate({
        age: '35',
      });
    } catch (error) {
      expect(error).toMatchSnapshot();
    }
  });
});

