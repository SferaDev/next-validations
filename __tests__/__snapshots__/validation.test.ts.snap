// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FastestValidator Validation > should create resolve base on schema 1`] = `
{
  "validate": [Function],
}
`;

exports[`FastestValidator Validation > should return false 1`] = `[Error: Validation failed]`;

exports[`Yoi Validation > should return false 1`] = `[ValidationError: "email" is required]`;

exports[`Yup Validation > should create yup resolve base on schema 1`] = `
{
  "validate": [Function],
}
`;

exports[`Yup Validation > should return false 1`] = `[ValidationError: age is a required field]`;

exports[`Zod Validation > should return false 1`] = `
[ZodError: [
  {
    "code": "too_small",
    "minimum": 8,
    "type": "string",
    "inclusive": true,
    "exact": false,
    "message": "String must contain at least 8 character(s)",
    "path": [
      "username"
    ]
  }
]]
`;
